src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+
+At any given point, there are at most levels/2 compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
+* The manifest is compacted at the end of every beat.
+* The compactions' output tables are not visible until the compaction has finished.
+
+1. (first) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
+    * Start even level compactions if there's any tables to compact.
+
+2. (second) up beat of the bar:
+    * Finish ticking running even-level compactions.
+    * Assert on callback completion that all compactions are complete.
+
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
+    * Compact the immutable table if it contains any sorted values (it could be empty).
+
+4. (fourth) last beat of the bar:
+    * Finish ticking running odd-level and immutable table compactions.
+    * Assert on callback completion that all compactions are complete.
+    * Assert on callback completion that no level's visible table count overflows.
+    * Flush, clear, and sort mutable table values into immutable table for next bar.
+
+
+## Snapshots
+### Snapshots and Compaction
+
+Consider the half-bar compaction beginning at op=`X` (`12`), with `lsm_batch_multiple=M` (`8`).
+Each half-bar contains `N=M/2` (`4`) beats. The next half-bar begins at `Y=X+N` (`16`).
+
+During the compaction `X` (op=`X…Y-1`; `12…15`), each commit prefetches from the snapshot equal to
+the first op of the compaction. As shown, they continue to query the old (input) tables.
+
+During the compaction `X`:
+- `snapshot_max` of each input table is truncated to `Y-1` (`15`).
+- `snapshot_min` of each output table is initialized to `Y` (`16`).
+
+```
+  0   4   8  12  16  20  24  (op, snapshot)
+  ┼───┬───┼───┬───┼───┬───┼
+              ####
+  ····────────X────────····  (input  tables, before compaction)
+  ····────────────           (input  tables,  after compaction)
+                  Y────····  (output tables,  after compaction)
+```
+
+Beginning from the next op after the compaction (`Y`; `16`):
+- The output tables of the above compaction `X` are visible.
+- The input tables of the above compaction `X` are invisible.
+- Therefore, it will lookup from the output tables, but ignore the input tables.
+- Commits must not prefetch from the output tables of `X` before the compaction half-bar
```

Comment:
'before the compaction half-bar has finished' = before Y?

Reply:
Not quite, but it's subtle. I'll clarify that to:
```
- Callers must not query from the output tables of `X` before the compaction half-bar has finished
  (i.e. before the end of beat `Y-1` (`15`)), since those tables are incomplete.
```

I also generalize from prefetching to more general queries, since from an abstraction-boundary point-of-view the `Tree` doesn't know anything about prefetches, just lookups.